File: packages/typescript/src/adapters/graphql.ts
/**
 * Apollo Server plugin adapter for Halt rate limiting.
 *
 * Usage:
 * const plugin = haltApolloPlugin({ limiter });
 * const server = new ApolloServer({ typeDefs, resolvers, plugins: [plugin] });
 */

import { RateLimiter } from '../core/limiter';

export interface ApolloHaltOptions {
    limiter: RateLimiter;
    /** Optional function to extract a resource name for per-field limits */
    resourceExtractor?: (request: any) => string | null;
}

export function haltApolloPlugin(options: ApolloHaltOptions) {
    const { limiter, resourceExtractor } = options;

    return {
        async requestDidStart() {
            return {
                async didResolveOperation(requestContext: any) {
                    try {
                        // Provide request object compatible with limiter.extractors
                        const req = requestContext.request?.http || {};
                        if (resourceExtractor) {
                            // Allow users to attach custom resource info
                            (req as any).resource = resourceExtractor(requestContext.request);
                        }

                        const decision = await limiter.check(req);

                        if (!decision.allowed) {
                            // Short-circuit the request by adding an error
                            requestContext.response = {
                                http: {
                                    status: 429,
                                    headers: {
                                        ...requestContext.response?.http?.headers,
                                        'Retry-After': String(decision.retryAfter ?? 0),
                                    },
                                },
                                errors: [
                                    new Error('rate_limit_exceeded: Too many requests. Please try again later.'),
                                ],
                            };
                        } else {
                            // Attach rate limit headers for downstream
                            requestContext.response = requestContext.response || {};
                            requestContext.response.http = requestContext.response.http || {};
                            requestContext.response.http.headers = {
                                ...requestContext.response.http.headers,
                                'RateLimit-Limit': String(decision.limit),
                                'RateLimit-Remaining': String(Math.max(0, decision.remaining)),
                                'RateLimit-Reset': String(decision.resetAt),
                            };
                        }
                    } catch (err) {
                        // Don't block on errors from limiter; surface as server error
                        requestContext.errors = requestContext.errors || [];
                        requestContext.errors.push(err as Error);
                    }
                },
            };
        },
    };
}

File: packages/python/halt/adapters/graphql.py
"""Graphene middleware adapter for Halt rate limiting.

Usage:
from halt.adapters.graphql import HaltGrapheneMiddleware

middleware = [HaltGrapheneMiddleware(limiter)]
schema = graphene.Schema(query=Query, middleware=middleware)
"""
from typing import Any, Optional


class HaltGrapheneMiddleware:
    def __init__(self, limiter: Any, resource_extractor: Optional[Any] = None):
        self.limiter = limiter
        self.resource_extractor = resource_extractor

    def resolve(self, next_, root, info, **kwargs):
        """Graphene middleware entrypoint. Runs before resolver execution."""
        # Build a request-like object the limiter understands
        request = getattr(info.context, "request", info.context)

        if self.resource_extractor:
            setattr(request, "resource", self.resource_extractor(info))

        # Limiter may be sync; handle both
        try:
            decision = self.limiter.check(request)
        except TypeError:
            # Async resolver - call coroutine
            import asyncio

            decision = asyncio.get_event_loop().run_until_complete(self.limiter.check(request))

        if not decision.allowed:
            # Raise an error to short-circuit resolver execution
            raise Exception("rate_limit_exceeded: Too many requests. Please try again later.")

        # Optionally attach headers on context for HTTP frameworks
        if hasattr(info.context, "response"):
            try:
                headers = getattr(info.context, "response").headers
                headers["RateLimit-Limit"] = str(decision.limit)
                headers["RateLimit-Remaining"] = str(max(0, decision.remaining))
                headers["RateLimit-Reset"] = str(decision.reset_at if hasattr(decision, "reset_at") else decision.resetAt)
            except Exception:
                pass

        return next_(root, info, **kwargs)
